File: 1.8.9-Forge/src/main/java/net/ccbluex/liquidbounce/ui/client/hud/element/elements/Image.java
package net.ccbluex.liquidbounce.ui.client.hud.element.elements;

import com.google.gson.JsonElement;
import net.ccbluex.liquidbounce.ui.client.hud.GuiHudDesigner;
import net.ccbluex.liquidbounce.ui.client.hud.element.Element;
import net.ccbluex.liquidbounce.ui.client.hud.element.ElementInfo;
import net.ccbluex.liquidbounce.utils.misc.RandomUtils;
import net.ccbluex.liquidbounce.utils.render.RenderUtils;
import net.ccbluex.liquidbounce.value.TextValue;
import net.minecraft.client.renderer.texture.DynamicTexture;
import net.minecraft.util.ResourceLocation;

import javax.imageio.ImageIO;
import java.awt.image.BufferedImage;
import java.io.ByteArrayInputStream;
import java.io.File;
import java.nio.file.Files;
import java.util.Base64;

/**
 * LiquidBounce Hacked Client
 * A minecraft forge injection client using Mixin
 *
 * @game Minecraft
 * @author CCBlueX
 */
@ElementInfo(name = "Image")
public class Image extends Element {

    private final TextValue image = new TextValue("Image", "") {

        @Override
        public void fromJson(JsonElement element) {
            super.fromJson(element);

            if(image.get().isEmpty()) return;

            setImage(image.get());
        }

        @Override
        protected void onChanged(final String oldValue, final String newValue) {
            if(image.get().isEmpty()) return;

            setImage(image.get());
        }
    };

    private final ResourceLocation resourceLocation = new ResourceLocation(RandomUtils.randomNumber(128));

    private int width = 64;
    private int height = 64;

    @Override
    public void drawElement() {
        final int[] location = getLocationFromFacing();

        RenderUtils.drawImage(resourceLocation, location[0], location[1], width / 2, height / 2);

        if (mc.currentScreen instanceof GuiHudDesigner)
            RenderUtils.drawBorderedRect(location[0], location[1], location[0] + width / 2, location[1] + height / 2, 3, Integer.MIN_VALUE, 0);
    }

    @Override
    public void destroyElement() {

    }

    @Override
    public void updateElement() {

    }

    @Override
    public void handleMouseClick(int mouseX, int mouseY, int mouseButton) {
    }

    @Override
    public void handleKey(char c, int keyCode) {
    }

    @Override
    public boolean isMouseOverElement(int mouseX, int mouseY) {
        final int[] location = getLocationFromFacing();

        return mouseX >= location[0] && mouseY >= location[1] && mouseX <= location[0] + width && mouseY <= location[1] + height;
    }

    private Image setImage(final String image) {
        try {
            this.image.changeValue(image);

            final ByteArrayInputStream byteArrayInputStream = new ByteArrayInputStream(Base64.getDecoder().decode(image));
            final BufferedImage bufferedImage = ImageIO.read(byteArrayInputStream);
            byteArrayInputStream.close();

            width = bufferedImage.getWidth();
            height = bufferedImage.getHeight();

            mc.getTextureManager().loadTexture(resourceLocation, new DynamicTexture(bufferedImage));
        }catch(final Exception e) {
            e.printStackTrace();
        }

        return this;
    }

    public Image setImage(final File image) {
        try {
            setImage(Base64.getEncoder().encodeToString(Files.readAllBytes(image.toPath())));
        }catch(final Exception e) {
            e.printStackTrace();
        }

        return this;
    }
}

File: src/main/java/net/ccbluex/liquidbounce/ui/client/hud/element/elements/Image.kt
/*
 * LiquidBounce Hacked Client
 * A free open source mixin-based injection hacked client for Minecraft using Minecraft Forge.
 * https://github.com/CCBlueX/LiquidBounce/
 */
package net.ccbluex.liquidbounce.ui.client.hud.element.elements

import com.google.gson.JsonElement
import net.ccbluex.liquidbounce.ui.client.hud.element.Border
import net.ccbluex.liquidbounce.ui.client.hud.element.Element
import net.ccbluex.liquidbounce.ui.client.hud.element.ElementInfo
import net.ccbluex.liquidbounce.utils.misc.MiscUtils
import net.ccbluex.liquidbounce.utils.misc.RandomUtils
import net.ccbluex.liquidbounce.utils.render.RenderUtils
import net.ccbluex.liquidbounce.value.TextValue
import net.minecraft.client.renderer.texture.DynamicTexture
import net.minecraft.util.ResourceLocation
import java.io.ByteArrayInputStream
import java.io.File
import java.nio.file.Files
import java.util.*
import javax.imageio.ImageIO


/**
 * CustomHUD image element
 *
 * Draw custom image
 */
@ElementInfo(name = "Image")
class Image : Element() {

    companion object {

        /**
         * Create default element
         */
        fun default(): Image {
            val image = Image()

            image.x = 0.0
            image.y = 0.0

            return image
        }

    }

    private val image: TextValue = object : TextValue("Image", "") {

        override fun fromJson(element: JsonElement) {
            super.fromJson(element)

            if (get().isEmpty())
                return

            setImage(get())
        }

        override fun onChanged(oldValue: String, newValue: String) {
            if (get().isEmpty())
                return

            setImage(get())
        }

    }

    private val resourceLocation = ResourceLocation(RandomUtils.randomNumber(128))
    private var width = 64
    private var height = 64

    /**
     * Draw element
     */
    override fun drawElement(): Border {
        RenderUtils.drawImage(resourceLocation, 0, 0, width / 2, height / 2)

        return Border(0F, 0F, width / 2F, height / 2F)
    }

    override fun createElement(): Boolean {
        val file = MiscUtils.openFileChooser() ?: return false

        if (!file.exists()) {
            MiscUtils.showErrorPopup("Error", "The file does not exist.")
            return false
        }

        if (file.isDirectory) {
            MiscUtils.showErrorPopup("Error", "The file is a directory.")
            return false
        }

        setImage(file)
        return true
    }

    private fun setImage(image: String): Image {
        try {
            this.image.changeValue(image)

            val byteArrayInputStream = ByteArrayInputStream(Base64.getDecoder().decode(image))
            val bufferedImage = ImageIO.read(byteArrayInputStream)
            byteArrayInputStream.close()

            width = bufferedImage.width
            height = bufferedImage.height

            mc.textureManager.loadTexture(resourceLocation, DynamicTexture(bufferedImage))
        } catch (e: Exception) {
            e.printStackTrace()
        }
        return this
    }

    fun setImage(image: File): Image {
        try {
            setImage(Base64.getEncoder().encodeToString(Files.readAllBytes(image.toPath())))
        } catch (e: Exception) {
            e.printStackTrace()
        }

        return this
    }

}